File: src/commands/courses.ts
import { getBaseDir, stripHtmlTags, formatTimestamp } from "../lib/utils.ts";
import { Command } from "commander";
import type { Logger, OutputFormat } from "../lib/types.ts";
import { getEnrolledCoursesApi } from "../lib/moodle.ts";
import { createLogger } from "../lib/logger.ts";
import { loadWsToken } from "../lib/token.ts";
import { formatAndOutput } from "../index.ts";
import path from "node:path";
import fs from "node:fs";

export function registerCoursesCommand(program: Command): void {
  const coursesCmd = program.command("courses");
  coursesCmd.description("Course operations");

  // Helper to get output format from global or local options
  function getOutputFormat(command: any): OutputFormat {
    const opts = command.optsWithGlobals();
    return (opts.output as OutputFormat) || "json";
  }

  // Pure API context - no browser required (fast!)
  async function createApiContext(options: { verbose?: boolean; headed?: boolean }, command?: any): Promise<{
    log: Logger;
    session: { wsToken: string; moodleBaseUrl: string };
  } | null> {
    const opts = command?.optsWithGlobals ? command.optsWithGlobals() : options;
    const outputFormat = getOutputFormat(command || { optsWithGlobals: () => ({ output: "json" }) });
    const silent = outputFormat === "json" && !opts.verbose;
    const log = createLogger(opts.verbose, silent, outputFormat);

    const baseDir = getBaseDir();
    const sessionPath = path.resolve(baseDir, ".auth", "storage-state.json");

    // Check if session exists
    if (!fs.existsSync(sessionPath)) {
      console.error("未找到登入 session。請先執行 'openape login' 進行登入。");
      log.info(`Session 預期位置: ${sessionPath}`);
      return null;
    }

    // Try to load WS token
    const wsToken = loadWsToken(sessionPath);
    if (!wsToken) {
      console.error("未找到 WS token。請先執行 'openape login' 進行登入。");
      return null;
    }

    return {
      log,
      session: {
        wsToken,
        moodleBaseUrl: "https://ilearning.cycu.edu.tw",
      },
    };
  }

  coursesCmd
    .command("list")
    .description("List enrolled courses")
    .option("--incomplete-only", "Show only incomplete courses")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .option("--level <type>", "Course level: in_progress (default) | past | future | all", "in_progress")
    .action(async (options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      // Map level to classification
      const classification = options.level === "all" ? undefined :
        options.level === "past" ? "past" :
          options.level === "future" ? "future" : "inprogress";

      const courses = await getEnrolledCoursesApi(apiContext.session, {
        classification,
      });

      let filteredCourses = courses;
      if (options.incompleteOnly) {
        filteredCourses = courses.filter(c => (c.progress ?? 0) < 100);
      }

      formatAndOutput(filteredCourses as unknown as Record<string, unknown>[], output, apiContext.log);
    });

  coursesCmd
    .command("info")
    .description("Show detailed course information")
    .argument("<course-id>", "Course ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);
      const course = courses.find(c => c.id === parseInt(courseId, 10));

      if (!course) {
        apiContext.log.error(`Course not found: ${courseId}`);
        process.exitCode = 1;
        return;
      }

      formatAndOutput(course as unknown as Record<string, unknown>, output, apiContext.log);
    });

  coursesCmd
    .command("progress")
    .description("Show course progress")
    .argument("<course-id>", "Course ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      const courses = await getEnrolledCoursesApi(apiContext.session);
      const course = courses.find(c => c.id === parseInt(courseId, 10));

      if (!course) {
        apiContext.log.error(`Course not found: ${courseId}`);
        process.exitCode = 1;
        return;
      }

      const progressData = {
        courseId: course.id,
        courseName: course.fullname,
        progress: course.progress ?? 0,
        startDate: course.startdate ? formatTimestamp(course.startdate) : null,
        endDate: course.enddate ? formatTimestamp(course.enddate) : null,
      };

      formatAndOutput(progressData as unknown as Record<string, unknown>, output, apiContext.log);
    });

  // Helper function to fetch syllabus from CMAP using GWT-RPC API
  async function fetchSyllabus(shortname: string): Promise<Record<string, unknown> | null> {
    try {
      const parts = shortname.split("_");
      if (parts.length < 2) {
        return { error: "Invalid course shortname format" };
      }

      const [yearTerm, opCode] = parts;

      // Build GWT-RPC request body
      // Format: 7|0|8|<base_url>|<permutation>|<service>|<method>|<param_types>|<params>...|1|2|3|4|3|5|5|5|6|7|8|
      const gwtBody = `7|0|8|https://cmap.cycu.edu.tw:8443/Syllabus/syllabus/|339796D6E7B561A6465F5E9B5F4943FA|com.sanfong.syllabus.shared.SyllabusClientService|findClassTargetByYearAndOpCode|java.lang.String/2004016611|${yearTerm}|${opCode}|zh_TW|1|2|3|4|3|5|5|5|6|7|8|`;

      const response = await fetch("https://cmap.cycu.edu.tw:8443/Syllabus/syllabus/syllabusClientService", {
        method: "POST",
        headers: {
          "X-GWT-Permutation": "339796D6E7B561A6465F5E9B5F4943FA",
          "Accept": "text/x-gwt-rpc, */*; q=0.01",
          "Content-Type": "text/x-gwt-rpc; charset=UTF-8",
        },
        body: gwtBody,
      });

      if (!response.ok) {
        return { error: `HTTP ${response.status}`, url: "https://cmap.cycu.edu.tw:8443/Syllabus/syllabus/syllabusClientService" };
      }

      const rawText = await response.text();

      // GWT-RPC response format: //OK[...data...]
      if (!rawText.startsWith("//OK")) {
        return { error: "Invalid GWT-RPC response", rawResponse: rawText.slice(0, 200) };
      }

      // Extract the JSON array part from the GWT response
      // Response format: //OK[data1,data2,...]
      const content = rawText.slice(4); // Remove "//OK"

      // Parse the GWT string table - GWT uses a special format where strings are escaped
      // Format: ["string1","string2",...] or [123,"string2",...]
      const stringTable: string[] = [];

      // Simple parser for GWT string table
      let current = "";
      let inString = false;
      let escaped = false;

      for (let i = 0; i < content.length; i++) {
        const char = content[i];

        if (escaped) {
          // Handle escape sequences
          switch (char) {
            case 'n':
              current += '\n';
              break;
            case 'r':
              current += '\r';
              break;
            case 't':
              current += '\t';
              break;
            case '"':
              current += '"';
              break;
            case '\\':
              current += '\\';
              break;
            case '0':
              current += '\0';
              break;
            default:
              // Unknown escape, just append the char
              current += char;
          }
          escaped = false;
          continue;
        }

        if (char === "\\") {
          escaped = true;
          continue;
        }

        if (char === '"') {
          inString = !inString;
          if (!inString && current.length > 0) {
            stringTable.push(current);
            current = "";
          }
          continue;
        }

        if (inString) {
          current += char;
        }
      }

      // Parse schedule from string table
      // Strategy: Find week numbers (1-18), extract title (previous field) and date
      const schedule: Array<{ week: string; date: string; title: string }> = [];
      const datePattern = /^\d{4}-\d{2}-\d{2}$/;

      // Track processed indices to avoid duplicates
      const processedIndices = new Set<number>();

      for (let i = 0; i < stringTable.length; i++) {
        const s = stringTable[i];

        // Look for week numbers (1-18)
        if (/^[1-9]$|^1[0-8]$/.test(s) && !processedIndices.has(i)) {
          const week = s;
          let date = "";
          let title = "";

          // Previous field is the title
          if (i - 1 >= 0 && !processedIndices.has(i - 1)) {
            title = stringTable[i - 1];
          }

          // For week 1 & 2, find date before the week number
          // For week 18, also look before (last week has no "next week")
          // For other weeks (3-17), next field is next week's date
          if (week === "1" || week === "2" || week === "18") {
            // Look backwards for date pattern (search further back for week 18)
            const maxLookback = week === "18" ? 15 : 6;
            for (let j = i - 1; j >= Math.max(0, i - maxLookback); j--) {
              if (datePattern.test(stringTable[j]) && !processedIndices.has(j)) {
                date = stringTable[j];
                processedIndices.add(j);
                break;
              }
            }
          } else {
            // Week 3-17: look for next week number, then get date before it
            for (let j = i + 1; j < Math.min(i + 10, stringTable.length); j++) {
              if (/^[1-9]$|^1[0-8]$/.test(stringTable[j]) && !processedIndices.has(j)) {
                // Found next week, look before it for date
                for (let k = j - 1; k >= Math.max(0, j - 6); k--) {
                  if (datePattern.test(stringTable[k]) && !processedIndices.has(k)) {
                    date = stringTable[k];
                    processedIndices.add(k);
                    break;
                  }
                }
                break;
              }
            }
          }

          // Clean up title
          title = title.trim()
            .replace(/[\r\n]+/g, ' ')
            .replace(/,+$/, '')
            .trim()
            .slice(0, 200);

          // Only add if we have title (date is optional, will be inferred if missing)
          if (title.length > 1) {
            // If no date found, try to infer from the last added date
            if (date.length === 0 && schedule.length > 0) {
              const lastEntry = schedule[schedule.length - 1];
              const lastDate = new Date(lastEntry.date);
              const nextDate = new Date(lastDate.getTime() + 7 * 24 * 60 * 60 * 1000);
              const year = nextDate.getFullYear();
              const month = String(nextDate.getMonth() + 1).padStart(2, "0");
              const day = String(nextDate.getDate()).padStart(2, "0");
              date = `${year}-${month}-${day}`;
            }

            schedule.push({
              week,
              date,
              title,
            });
          }

          processedIndices.add(i);
        }
      }

      // Sort by date to maintain order
      schedule.sort((a, b) => a.date.localeCompare(b.date));

      // Extract course info from string table
      const result: Record<string, unknown> = {
        yearTerm,
        opCode,
        url: `https://cmap.cycu.edu.tw:8443/Syllabus/CoursePreview.html?yearTerm=${yearTerm}&opCode=${opCode}&locale=zh_TW`,
        schedule,
      };

      // Try to find instructor (look for common patterns)
      for (let i = 0; i < stringTable.length; i++) {
        const s = stringTable[i];
        if (s.includes("教授") || s.includes("老師") || s.includes("教師") || s.includes("Instructor")) {
          result.instructor = s;
          break;
        }
      }

      return result;
    } catch (e) {
      return { error: e instanceof Error ? e.message : String(e) };
    }
  }

  coursesCmd
    .command("syllabus")
    .description("Show course syllabus (from CMAP)")
    .argument("<course-id>", "Course ID")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      try {
        const courses = await getEnrolledCoursesApi(apiContext.session);
        const course = courses.find(c => c.id === parseInt(courseId, 10));

        if (!course) {
          apiContext.log.error(`Course not found: ${courseId}`);
          process.exitCode = 1;
          return;
        }

        // Fetch syllabus from CMAP
        const syllabus = await fetchSyllabus(course.shortname);

        if (!syllabus) {
          apiContext.log.warn(`Syllabus not found for course: ${course.shortname}`);
          // Return course info at least
          formatAndOutput({
            courseId: course.id,
            shortname: course.shortname,
            fullname: course.fullname,
            note: "Syllabus not available from CMAP",
          } as unknown as Record<string, unknown>, output, apiContext.log);
          return;
        }

        // Combine course info with syllabus
        const result = {
          courseId: course.id,
          shortname: course.shortname,
          fullname: course.fullname,
          ...syllabus,
        };

        formatAndOutput(result as unknown as Record<string, unknown>, output, apiContext.log);
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        apiContext.log.error(`Error fetching syllabus: ${msg}`);
        process.exitCode = 1;
      }
    });
}

File: src/commands/courses.rs
use anyhow::Result;
use crate::Cli;
use crate::moodle::course::get_enrolled_courses_api;
use crate::output::format_and_output;
use crate::utils::format_moodle_date;
use super::{ApiCtx, level_to_classification};

pub async fn run(cmd: &crate::CoursesCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli)?;

    match cmd {
        crate::CoursesCommands::List { incomplete_only, level } => {
            let classification = level_to_classification(*level);
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, classification).await?;

            let filtered: Vec<_> = if *incomplete_only {
                courses.into_iter().filter(|c| c.progress.unwrap_or(0) < 100).collect()
            } else {
                courses
            };

            let items: Vec<serde_json::Value> = filtered.iter()
                .map(|c| serde_json::json!({
                    "id": c.id,
                    "fullname": c.fullname,
                    "shortname": c.shortname,
                    "progress": c.progress,
                    "startdate": format_moodle_date(c.startdate),
                    "enddate": format_moodle_date(c.enddate),
                }))
                .collect();

            format_and_output(&items, ctx.output, None);
        }

        crate::CoursesCommands::Info { course_id } => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "all").await?;
            let course = courses.iter().find(|c| c.id == *course_id)
                .ok_or_else(|| anyhow::anyhow!("Course not found: {}", course_id))?;

            let item = serde_json::to_value(course)?;
            format_and_output(&[item], ctx.output, None);
        }

        crate::CoursesCommands::Progress { course_id } => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "all").await?;
            let course = courses.iter().find(|c| c.id == *course_id)
                .ok_or_else(|| anyhow::anyhow!("Course not found: {}", course_id))?;

            let item = serde_json::json!({
                "courseId": course.id,
                "courseName": course.fullname,
                "progress": course.progress.unwrap_or(0),
                "startDate": format_moodle_date(course.startdate),
                "endDate": format_moodle_date(course.enddate),
            });

            format_and_output(&[item], ctx.output, None);
        }

        crate::CoursesCommands::Syllabus { course_id } => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "all").await?;
            let course = courses.iter().find(|c| c.id == *course_id)
                .ok_or_else(|| anyhow::anyhow!("Course not found: {}", course_id))?;

            let syllabus = fetch_syllabus(&ctx.client, &course.shortname).await;

            let mut result = serde_json::json!({
                "courseId": course.id,
                "shortname": course.shortname,
                "fullname": course.fullname,
            });

            match syllabus {
                Some(s) => {
                    if let serde_json::Value::Object(ref mut map) = result {
                        if let serde_json::Value::Object(extra) = s {
                            map.extend(extra);
                        }
                    }
                }
                None => {
                    ctx.log.warn(&format!("Syllabus not found for course: {}", course.shortname));
                    if let serde_json::Value::Object(ref mut map) = result {
                        map.insert("note".to_string(), serde_json::json!("Syllabus not available from CMAP"));
                    }
                }
            }

            format_and_output(&[result], ctx.output, None);
        }
    }

    Ok(())
}

/// Fetch course syllabus from CYCU CMAP GWT-RPC endpoint.
async fn fetch_syllabus(client: &reqwest::Client, shortname: &str) -> Option<serde_json::Value> {
    let parts: Vec<&str> = shortname.splitn(2, '_').collect();
    if parts.len() < 2 {
        return None;
    }
    let (year_term, op_code) = (parts[0], parts[1]);

    let gwt_body = format!(
        "7|0|8|https://cmap.cycu.edu.tw:8443/Syllabus/syllabus/|339796D6E7B561A6465F5E9B5F4943FA|\
        com.sanfong.syllabus.shared.SyllabusClientService|findClassTargetByYearAndOpCode|\
        java.lang.String/2004016611|{}|{}|zh_TW|1|2|3|4|3|5|5|5|6|7|8|",
        year_term, op_code
    );

    let resp = client
        .post("https://cmap.cycu.edu.tw:8443/Syllabus/syllabus/syllabusClientService")
        .header("X-GWT-Permutation", "339796D6E7B561A6465F5E9B5F4943FA")
        .header("Accept", "text/x-gwt-rpc, */*; q=0.01")
        .header("Content-Type", "text/x-gwt-rpc; charset=UTF-8")
        .body(gwt_body)
        .send()
        .await
        .ok()?;

    if !resp.status().is_success() {
        return Some(serde_json::json!({
            "error": format!("HTTP {}", resp.status()),
        }));
    }

    let raw = resp.text().await.ok()?;
    if !raw.starts_with("//OK") {
        return Some(serde_json::json!({
            "error": "Invalid GWT-RPC response",
            "rawResponse": &raw[..200.min(raw.len())],
        }));
    }

    // Parse GWT string table
    let content = &raw[4..];
    let mut string_table: Vec<String> = Vec::new();
    let mut current = String::new();
    let mut in_string = false;
    let mut escaped = false;

    for ch in content.chars() {
        if escaped {
            match ch {
                'n' => current.push('\n'),
                'r' => current.push('\r'),
                't' => current.push('\t'),
                '"' => current.push('"'),
                '\\' => current.push('\\'),
                '0' => current.push('\0'),
                other => current.push(other),
            }
            escaped = false;
            continue;
        }
        if ch == '\\' { escaped = true; continue; }
        if ch == '"' {
            in_string = !in_string;
            if !in_string && !current.is_empty() {
                string_table.push(std::mem::take(&mut current));
            }
            continue;
        }
        if in_string { current.push(ch); }
    }

    // Extract schedule from string table (week numbers 1-18)
    let date_pattern = regex::Regex::new(r"^\d{4}-\d{2}-\d{2}$").ok()?;
    let week_pattern = regex::Regex::new(r"^[1-9]$|^1[0-8]$").ok()?;

    let mut schedule: Vec<serde_json::Value> = Vec::new();
    let mut processed: std::collections::HashSet<usize> = std::collections::HashSet::new();

    for i in 0..string_table.len() {
        let s = &string_table[i];
        if !week_pattern.is_match(s) || processed.contains(&i) { continue; }

        let week = s.clone();
        let mut date = String::new();
        let mut title = String::new();

        if i > 0 && !processed.contains(&(i - 1)) {
            title = string_table[i - 1].clone();
        }

        let is_edge = week == "1" || week == "2" || week == "18";
        if is_edge {
            let max_lookback = if week == "18" { 15 } else { 6 };
            for j in (i.saturating_sub(max_lookback)..i).rev() {
                if date_pattern.is_match(&string_table[j]) && !processed.contains(&j) {
                    date = string_table[j].clone();
                    processed.insert(j);
                    break;
                }
            }
        } else {
            for j in (i + 1)..string_table.len().min(i + 10) {
                if week_pattern.is_match(&string_table[j]) && !processed.contains(&j) {
                    for k in (j.saturating_sub(6)..j).rev() {
                        if date_pattern.is_match(&string_table[k]) && !processed.contains(&k) {
                            date = string_table[k].clone();
                            processed.insert(k);
                            break;
                        }
                    }
                    break;
                }
            }
        }

        // Clean title
        let title = title.trim()
            .replace(['\r', '\n'], " ")
            .trim_end_matches(',')
            .trim()
            .chars()
            .take(200)
            .collect::<String>();

        if title.len() > 1 {
            // Infer date if missing
            let date = if date.is_empty() {
                if let Some(last) = schedule.last() {
                    if let Some(last_date) = last.get("date").and_then(|d| d.as_str()) {
                        infer_next_week_date(last_date)
                    } else { date }
                } else { date }
            } else { date };

            schedule.push(serde_json::json!({ "week": week, "date": date, "title": title }));
        }

        processed.insert(i);
    }

    schedule.sort_by(|a, b| {
        let da = a.get("date").and_then(|v| v.as_str()).unwrap_or("");
        let db = b.get("date").and_then(|v| v.as_str()).unwrap_or("");
        da.cmp(db)
    });

    let mut result = serde_json::json!({
        "yearTerm": year_term,
        "opCode": op_code,
        "url": format!("https://cmap.cycu.edu.tw:8443/Syllabus/CoursePreview.html?yearTerm={}&opCode={}&locale=zh_TW", year_term, op_code),
        "schedule": schedule,
    });

    // Try to find instructor
    for s in &string_table {
        if s.contains("教授") || s.contains("老師") || s.contains("教師") || s.contains("Instructor") {
            if let serde_json::Value::Object(ref mut map) = result {
                map.insert("instructor".to_string(), serde_json::json!(s));
            }
            break;
        }
    }

    Some(result)
}

fn infer_next_week_date(last_date: &str) -> String {
    use std::str::FromStr;
    let parts: Vec<&str> = last_date.splitn(3, '-').collect();
    if parts.len() != 3 { return String::new(); }
    let (y, m, d) = (
        i32::from_str(parts[0]).unwrap_or(2024),
        u32::from_str(parts[1]).unwrap_or(1),
        u32::from_str(parts[2]).unwrap_or(1),
    );
    // Simple +7 days calculation
    let days_in_month = [0u32, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    let leap = y % 4 == 0 && (y % 100 != 0 || y % 400 == 0);
    let feb_days = if leap { 29u32 } else { 28 };
    let dim = if m == 2 { feb_days } else { days_in_month[m as usize] };
    let (ny, nm, nd) = if d + 7 > dim {
        let nd = d + 7 - dim;
        if m == 12 { (y + 1, 1, nd) } else { (y, m + 1, nd) }
    } else {
        (y, m, d + 7)
    };
    format!("{:04}-{:02}-{:02}", ny, nm, nd)
}
